cloud/google/cmd/gce-machine-controller/app/controller.go

```diff
@@ -0,0 +1,153 @@
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/cluster-api/cloud/google"
+	"sigs.k8s.io/cluster-api/cloud/google/cmd/gce-machine-controller/app/options"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	clusterapiclientsetscheme "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/scheme"
+	"sigs.k8s.io/cluster-api/pkg/controller/config"
+	"sigs.k8s.io/cluster-api/pkg/controller/machine"
+	"sigs.k8s.io/cluster-api/pkg/controller/sharedinformers"
+)
+
+const (
+	gceMachineControllerName = "gce-machine-controller"
+)
+
+func StartMachineController(server *options.MachineControllerServer, shutdown <-chan struct{}) {
+	config, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Fatalf("Could not create Config for talking to the apiserver: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Could not create client for talking to the apiserver: %v", err)
+	}
+
+	actuator, err := google.NewMachineActuator(server.KubeadmToken, client.ClusterV1alpha1().Machines(corev1.NamespaceDefault), server.MachineSetupConfigsPath)
+	if err != nil {
+		glog.Fatalf("Could not create Google machine actuator: %v", err)
+	}
+
+	si := sharedinformers.NewSharedInformers(config, shutdown)
+	// If this doesn't compile, the code generator probably
+	// overwrote the customized NewMachineController function.
+	c := machine.NewMachineController(config, si, actuator)
+	c.RunAsync(shutdown)
+
+	select {}
+}
+
+func Run(server *options.MachineControllerServer) error {
+	kubeConfig, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Errorf("Could not create Config for talking to the apiserver: %v", err)
+		return err
+	}
+
+	kubeClientControl, err := kubernetes.NewForConfig(
+		rest.AddUserAgent(kubeConfig, "machine-controller-manager"),
+	)
+	if err != nil {
+		glog.Errorf("Invalid API configuration for kubeconfig-control: %v", err)
+		return err
+	}
+
+	recorder, err := createRecorder(kubeClientControl)
+	if err != nil {
+		glog.Errorf("Could not create event recorder : %v", err)
+		return err
+	}
+
+	run := func(stop <-chan struct{}) {
+		StartMachineController(server, stop)
+		panic("unreachable")
+	}
+
+	leaderElectConfig := config.GetLeaderElectionConfig()
+	if !leaderElectConfig.LeaderElect {
+		run(make(<-chan (struct{})))
+		panic("unreachable")
+	}
+
+	// Identity used to distinguish between multiple machine controller instances.
+	id, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	leaderElectionClient := kubernetes.NewForConfigOrDie(rest.AddUserAgent(kubeConfig, "machine-leader-election"))
+
+	// Lock required for leader election
+	rl, err := resourcelock.New(
+		leaderElectConfig.ResourceLock,
+		"default",
```

Comment:
Isn't it better to run in the `kube-system` namespace? There are constant for those names in apimachinery - `metav1.NamespaceSystem`

Reply:
I have a PR out soon to make namespace configurable for controller to look up/reference resource. Right now everything is assumed to be in default namespace. For this PR, i changed it to metav1.NamespaceSystem as you suggested.